core/src/value/deserialize.rs

```diff
@@ -29,55 +50,286 @@ pub enum Error {
     InvalidNumberToken { value: u8, backtrace: Backtrace },
     #[snafu(display("Invalid time zone sign token: got '{}', but must be '+' or '-'", *value as char))]
     InvalidTimeZoneSignToken { value: u8, backtrace: Backtrace },
+    #[snafu(display(
+        "Could not parse incomplete value: first missing component: {:?}",
+        component
+    ))]
+    IncompleteValue {
+        component: DateComponent,
+        backtrace: Backtrace,
+    },
+    #[snafu(display("Range is zero"))]
+    ZeroRange { backtrace: Backtrace },
 }
 
 type Result<T, E = Error> = std::result::Result<T, E>;
 
+#[derive(Debug, PartialEq)]
+pub enum DateComponent {
+    Year,
+    Month,
+    Day,
+    Hour,
+    Minute,
+    Second,
+    Fraction,
+    UTCOffset,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum DatePrecision {
+    Year(u16),
+    Month(u16, u8),
+    Day(u16, u8, u8),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct PartialDate {
+    components: DatePrecision,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum DateTimePrecision {
+    Year(u16),
+    Month(u16, u8),
+    Day(u16, u8, u8),
+    Hour(u16, u8, u8, u8),
+    Minute(u16, u8, u8, u8, u8),
+    Second(u16, u8, u8, u8, u8, u8),
+    Fraction(u16, u8, u8, u8, u8, u8, u32),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct PartialDateTime {
+    components: DateTimePrecision,
+    // TODO probably not an option, will see ..
+    offset: Option<FixedOffset>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum TimePrecision {
+    Hour(u8),
+    Minute(u8, u8),
+    Second(u8, u8, u8),
+    Fraction(u8, u8, u8, u32),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct PartialTime {
+    components: TimePrecision,
+}
+
+/**
+ * This trait is implemented by partial precision
+ * Date, Time and DateTime structures.
+ * Trait method returns the last fully precise component of the value.
+ */
+pub trait Precision {
+    fn precision(&self) -> DateComponent;
+}
+
+impl Precision for PartialDate {
+    fn precision(&self) -> DateComponent {
+        match self.components {
+            DatePrecision::Year(..) => DateComponent::Year,
+            DatePrecision::Month(..) => DateComponent::Month,
+            DatePrecision::Day(..) => DateComponent::Day,
+        }
+    }
+}
+
+/**
+ * This trait is implemented by partial precision
+ * Date, Time and DateTime structures.
+ */
+pub trait Imprecision<U>
```

Comment:
This seems to call for a super-trait, rather than a constraint `Self: Precision`
```suggestion
pub trait Imprecision<U>: Precision
```

Better clarify what `U` is in this context. There may also be better names for these traits... `AsTemporalRange<T>`, maybe?